File: src/TrackForge/Core/ToolInstaller.cs
using System.IO.Compression;
using System.Net.Http;

namespace TrackForge.Core;

/// <summary>
/// Fetches yt-dlp and ffmpeg into TrackForge's own tools folder so the app works on
/// a clean machine without the user knowing what a PATH is.
///
/// These are downloaded rather than shipped inside the installer on purpose: ffmpeg
/// builds are GPL and redistributing them carries source-offer obligations, and
/// yt-dlp goes stale within weeks. Fetching from upstream sidesteps both.
/// </summary>
public static class ToolInstaller
{
    public const string YtDlpUrl =
        "https://github.com/yt-dlp/yt-dlp/releases/latest/download/yt-dlp.exe";

    // Gyan's "essentials" build: a stable URL that always points at the current release.
    public const string FfmpegZipUrl =
        "https://www.gyan.dev/ffmpeg/builds/ffmpeg-release-essentials.zip";

    /// <summary>%LOCALAPPDATA%\TrackForge\tools - writable without admin rights.</summary>
    public static string ToolsDirectory { get; } = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "TrackForge", "tools");

    public static string YtDlpExe => Path.Combine(ToolsDirectory, "yt-dlp.exe");
    public static string FfmpegExe => Path.Combine(ToolsDirectory, "ffmpeg.exe");
    public static string FfprobeExe => Path.Combine(ToolsDirectory, "ffprobe.exe");

    public static bool HasYtDlp => File.Exists(YtDlpExe);
    public static bool HasFfmpeg => File.Exists(FfmpegExe);

    private static HttpClient NewClient() =>
        new(new HttpClientHandler { AllowAutoRedirect = true })
        {
            Timeout = TimeSpan.FromMinutes(10),
            DefaultRequestHeaders = { { "User-Agent", "TrackForge/1.0" } },
        };

    public static async Task InstallYtDlpAsync(
        IProgress<(double percent, string message)>? progress, CancellationToken ct = default)
    {
        Directory.CreateDirectory(ToolsDirectory);
        progress?.Report((0, "Downloading yt-dlp"));

        var temp = YtDlpExe + ".part";
        using (var http = NewClient())
            await DownloadAsync(http, YtDlpUrl, temp, progress, "yt-dlp", ct).ConfigureAwait(false);

        if (File.Exists(YtDlpExe)) File.Delete(YtDlpExe);
        File.Move(temp, YtDlpExe);
        progress?.Report((100, "yt-dlp installed"));
    }

    public static async Task InstallFfmpegAsync(
        IProgress<(double percent, string message)>? progress, CancellationToken ct = default)
    {
        Directory.CreateDirectory(ToolsDirectory);
        var zipPath = Path.Combine(Path.GetTempPath(), "trackforge-ffmpeg.zip");

        using (var http = NewClient())
            await DownloadAsync(http, FfmpegZipUrl, zipPath, progress, "ffmpeg", ct).ConfigureAwait(false);

        progress?.Report((92, "Extracting ffmpeg"));

        // The zip nests everything under ffmpeg-<version>-essentials_build/bin/.
        await Task.Run(() =>
        {
            using var archive = ZipFile.OpenRead(zipPath);
            foreach (var wanted in new[] { "ffmpeg.exe", "ffprobe.exe" })
            {
                var entry = archive.Entries.FirstOrDefault(e =>
                    string.Equals(e.Name, wanted, StringComparison.OrdinalIgnoreCase));
                if (entry is null) continue;

                var target = Path.Combine(ToolsDirectory, wanted);
                if (File.Exists(target)) File.Delete(target);
                entry.ExtractToFile(target, overwrite: true);
            }
        }, ct).ConfigureAwait(false);

        try { File.Delete(zipPath); } catch { }

        if (!File.Exists(FfmpegExe))
            throw new InvalidOperationException("ffmpeg.exe was not found inside the download.");

        progress?.Report((100, "ffmpeg installed"));
    }

    private static async Task DownloadAsync(
        HttpClient http, string url, string destination,
        IProgress<(double, string)>? progress, string label, CancellationToken ct)
    {
        using var response = await http
            .GetAsync(url, HttpCompletionOption.ResponseHeadersRead, ct).ConfigureAwait(false);
        response.EnsureSuccessStatusCode();

        var total = response.Content.Headers.ContentLength ?? 0;
        Directory.CreateDirectory(Path.GetDirectoryName(destination)!);

        await using var source = await response.Content.ReadAsStreamAsync(ct).ConfigureAwait(false);
        await using var target = File.Create(destination);

        var buffer = new byte[81920];
        long read = 0;
        int n;
        while ((n = await source.ReadAsync(buffer, ct).ConfigureAwait(false)) > 0)
        {
            await target.WriteAsync(buffer.AsMemory(0, n), ct).ConfigureAwait(false);
            read += n;

            if (total > 0)
            {
                var percent = read * 90.0 / total;
                progress?.Report((percent, $"Downloading {label}  {read / 1048576.0:0.0} / {total / 1048576.0:0.0} MB"));
            }
            else
            {
                progress?.Report((0, $"Downloading {label}  {read / 1048576.0:0.0} MB"));
            }
        }
    }

    /// <summary>Pulls the newest yt-dlp over the top of the existing one.</summary>
    public static Task UpdateYtDlpAsync(
        IProgress<(double, string)>? progress, CancellationToken ct = default)
        => InstallYtDlpAsync(progress, ct);
}

File: macOS/Sources/ToolInstaller.swift
import Foundation

/// Fetches yt-dlp and ffmpeg into TrackForge's own tools folder so the app works
/// on a clean Mac without the user knowing what Homebrew is.
///
/// These are downloaded rather than shipped inside the app on purpose: ffmpeg
/// builds are GPL and redistributing them carries source-offer obligations, and
/// yt-dlp goes stale within weeks of a YouTube change. Fetching from upstream
/// sidesteps both, and gives the user a working update button.
enum ToolInstaller {

    /// The official standalone build — a universal binary with Python baked in,
    /// so there is nothing to install alongside it.
    static let ytDlpURL =
        "https://github.com/yt-dlp/yt-dlp/releases/latest/download/yt-dlp_macos"

    /// Per-architecture static ffmpeg. evermeet.cx ships a newer build but only
    /// for Intel; on Apple silicon that would quietly depend on Rosetta being
    /// installed, which it increasingly is not.
    static var ffmpegURL: String {
        let arch = isAppleSilicon ? "arm64" : "x64"
        return "https://github.com/eugeneware/ffmpeg-static/releases/latest/download/ffmpeg-darwin-\(arch)"
    }

    static var isAppleSilicon: Bool {
        var info = utsname()
        uname(&info)
        let machine = withUnsafeBytes(of: &info.machine) { raw -> String in
            String(cString: raw.baseAddress!.assumingMemoryBound(to: CChar.self))
        }
        return machine.hasPrefix("arm")
    }

    /// ~/Library/Application Support/TrackForge/tools — writable without admin rights.
    static let toolsDirectory = AppConfig.configDirectory
        .appendingPathComponent("tools", isDirectory: true)

    static var ytDlpPath: String { toolsDirectory.appendingPathComponent("yt-dlp").path }
    static var ffmpegPath: String { toolsDirectory.appendingPathComponent("ffmpeg").path }

    static var hasYtDlp: Bool { FileManager.default.isExecutableFile(atPath: ytDlpPath) }
    static var hasFfmpeg: Bool { FileManager.default.isExecutableFile(atPath: ffmpegPath) }

    struct Progress {
        var percent: Double
        var message: String
    }

    static func installYtDlp(onProgress: @escaping (Progress) -> Void) async throws {
        try await install(from: ytDlpURL, to: ytDlpPath, label: "yt-dlp", onProgress: onProgress)
    }

    static func installFfmpeg(onProgress: @escaping (Progress) -> Void) async throws {
        try await install(from: ffmpegURL, to: ffmpegPath, label: "ffmpeg", onProgress: onProgress)
    }

    /// Pulls the newest yt-dlp over the top of the existing one.
    static func updateYtDlp(onProgress: @escaping (Progress) -> Void) async throws {
        try await installYtDlp(onProgress: onProgress)
    }

    private static func install(
        from urlString: String, to destination: String, label: String,
        onProgress: @escaping (Progress) -> Void
    ) async throws {
        guard let url = URL(string: urlString) else {
            throw ToolError.message("That download address is not valid.")
        }

        try FileManager.default.createDirectory(
            at: toolsDirectory, withIntermediateDirectories: true)

        onProgress(Progress(percent: 0, message: "Downloading \(label)"))

        var request = URLRequest(url: url)
        request.setValue("TrackForge/1.0", forHTTPHeaderField: "User-Agent")
        request.timeoutInterval = 600

        let (temporary, response) = try await URLSession.shared.download(for: request)
        defer { try? FileManager.default.removeItem(at: temporary) }

        if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            throw ToolError.message("\(label) download failed (HTTP \(http.statusCode)).")
        }

        let size = (try? FileManager.default
            .attributesOfItem(atPath: temporary.path)[.size] as? NSNumber)??.int64Value ?? 0
        guard size > 1_000_000 else {
            throw ToolError.message("The \(label) download came back empty.")
        }

        onProgress(Progress(percent: 94, message: "Installing \(label)"))

        let target = URL(fileURLWithPath: destination)
        try? FileManager.default.removeItem(at: target)
        try FileManager.default.moveItem(at: temporary, to: target)

        try FileManager.default.setAttributes(
            [.posixPermissions: 0o755], ofItemAtPath: destination)

        // A file written by URLSession carries no quarantine flag, but clear it
        // anyway in case one is inherited, and give the binary an ad-hoc signature.
        // Apple silicon refuses to execute an unsigned arm64 binary outright.
        await shell("/usr/bin/xattr", ["-d", "com.apple.quarantine", destination])
        await shell("/usr/bin/codesign", ["--force", "--sign", "-", destination])

        onProgress(Progress(percent: 100, message: "\(label) installed"))
    }

    /// Best effort — neither of these failing is a reason to abandon the install.
    @discardableResult
    private static func shell(_ executable: String, _ arguments: [String]) async -> Bool {
        guard FileManager.default.isExecutableFile(atPath: executable) else { return false }
        let process = Process()
        process.executableURL = URL(fileURLWithPath: executable)
        process.arguments = arguments
        process.standardOutput = FileHandle.nullDevice
        process.standardError = FileHandle.nullDevice
        do {
            try process.run()
            process.waitUntilExit()
            return process.terminationStatus == 0
        } catch {
            return false
        }
    }
}

enum ToolError: LocalizedError {
    case message(String)
    var errorDescription: String? {
        if case .message(let text) = self { return text }
        return nil
    }
}
